src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java

```diff
@@ -222,22 +231,69 @@ public void buildGraph(Graph graph, HashMap<Class<?>, Object> extra) {
         assignMissingElevations(graph, edgesWithCalculatedElevations, extraElevation);
     }
 
+    /**
+     * A special extension of the Thread class so that a thread-specific coverage instance can be stored for each
+     * thread.
+     */
+    private class ElevationWorkerThread extends ForkJoinWorkerThread {
+        private Coverage threadSpecificCoverage;
+
+        /**
+         * Creates a ForkJoinWorkerThread operating in the given pool.
+         *
+         * @param pool the pool this thread works in
+         * @throws NullPointerException if pool is null
+         */
+        protected ElevationWorkerThread(ForkJoinPool pool) {
+            super(pool);
+        }
+
+        /**
+         * For unknown reasons, the interpolation of heights at coordinates is a synchronized method in the commonly
+         * used Interpolator2D class. Therefore, it is critical to use a dedicated Coverage instance for each thread to
+         * avoid other threads waiting for a lock to be released on the Coverage instance. This concurrent HashMap will
+         * store these thread-specific Coverage instances.
+         *
+         * @param swee An exemplar StreetWithElevationEdge to use to initialize a few calculations in the newly created
+         *             coverage instance.
+         * @return A thread-specific coverage instance that can be used without being locked from other threads.
+         */
+        public Coverage getThreadSpecificCoverageInstance (StreetWithElevationEdge swee) {
+            if (threadSpecificCoverage == null) {
+                // Synchronize the creation of the Thread-specific Coverage instances to avoid potential locks that
+                // could arise from downstream classes that have synchronized methods.
+                synchronized (gridCoverageFactory) {
+                    // Get a new Coverage instance from the module's ElevationGridCoverageFactory.
+                    threadSpecificCoverage = gridCoverageFactory.getGridCoverage();
+                    // The Coverage instance relies on some synchronized static methods shared across all threads that
+                    // can cause deadlocks if not fully initialized. Therefore, make a single request for the first
+                    // point on the edge to initialize these other items.
+                    Coordinate firstEdgeCoord =  swee.getGeometry().getCoordinates()[0];
+                    double[] dummy = new double[1];
+                    threadSpecificCoverage.evaluate(
+                        new DirectPosition2D(GeometryUtils.WGS84_XY, firstEdgeCoord.x, firstEdgeCoord.y),
+                        dummy
+                    );
+                }
```

Comment:
Do the call to the `evaluate` method need to be called with the first coordinate from the `swee` or just any coordinate? If any coordinate will do it, then you could create the `Coverage` in the constructor and make the  `threadSpecificCoverage` `final`. Also, I think the `synchronized` can be dropped in that case, because the creation of a new Thread should be thread-safe.

_Note! The above code is NOT thread-safe you check the `threadSpecificCoverage == null` outside of the synchronized block and assign it inside._

Reply:
I have pushed a change that will hopefully resolve the issues you brought up. The reasoning behind not immediately creating a coverage instance is that it can cause a lot of blocking in the beginning when the ForkJoinPool begins. When using the cached elevations, it is highly likely that most edges will already have a value present in the cache and therefore not even need a coverage instance. By lazy-creating the coverage instances other threads can continue to process edges that are available in the cache that don't need a coverage instance, otherwise, a lot of threads will be waiting for each other at the start until a coverage instance has been calculated for each thread.

Regarding the thread-safe assignment, I don't fully understand how it would not be thread-safe to assign the value in this way since the field would only be accessed by the thread doing the assignment and not by anything else. Nonetheless, I attempted to break out the assignment of the variable as suggested, so hopefully, this now makes it thread-safe.

Reply:
Sorry, I forgot the `threadSpecificCoverage` was inside the thread. It looks ok then. You could also have achieved this with a `ThreadLocal` but this good. The main point is to avoid every task to wait on the `synchronized (gridCoverageFactory)`.